bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java

```diff
@@ -3027,6 +3032,34 @@ private static boolean checkTupleCast(BValue sourceValue, BType targetType, List
         return true;
     }
 
+    private static boolean isAnydata(BType type) {
+        if (type.getTag() <= TypeTags.ANYDATA_TAG) {
+            return true;
+        }
+
+        switch (type.getTag()) {
+            case TypeTags.MAP_TAG:
+                return isAnydata(((BMapType) type).getConstrainedType());
+            case TypeTags.RECORD_TYPE_TAG:
+                BRecordType
+                        recordType = (BRecordType) type;
+                List<BType> fieldTypes = Arrays.stream(recordType.getFields())
+                        .map(BField::getFieldType)
+                        .collect(Collectors.toList());
+                return isAnydata(fieldTypes) && (recordType.sealed || isAnydata(recordType.restFieldType));
+            case TypeTags.UNION_TAG:
+                return isAnydata(((BUnionType) type).getMemberTypes());
+            case TypeTags.TUPLE_TAG:
+                return isAnydata(((BTupleType) type).getTupleTypes());
+            default:
+                return type.getTag() == TypeTags.ARRAY_TAG && isAnydata(((BArrayType) type).getElementType());
```

Comment:
IMO, it would be clearer to have a case for for ARRAY_TAG also, and return false for the default case or at the end.